pretyping/tacred.ml

```diff
@@ -268,7 +268,35 @@ let invert_name labs l {binder_name=na0} env sigma ref na =
    last one before the Fix if the latter is mutually defined *)
 
 let compute_consteval_direct env sigma ref =
-  let rec srec env n labs onlyproj c =
+  let get_value env eval =
+    match unsafe_reference_opt_value env sigma eval with
+    | None -> raise Redelimination
+    | Some c2 -> c2 in
+  let rec check_no_nosimpl env c =
+      let (c', l) = whd_betalet_stack env sigma c in
+      match EConstr.kind sigma c' with
+      | Const(ct, u) ->
+        begin
+          match ReductionBehaviour.get (GlobRef.ConstRef ct) with
+          | Some ReductionBehaviour.NeverUnfold -> raise Redelimination
+          | _ -> check_no_nosimpl env
+                      (applist (get_value env (EvalConst ct), l))
+        end
+      | Var id ->
+        check_no_nosimpl env (applist (get_value env (EvalVar id), l))
+      | Rel n ->
+        check_no_nosimpl env (applist (get_value env (EvalRel n), l))
+      | Evar ev ->
+        check_no_nosimpl env (applist (get_value env (EvalEvar ev), l))
+      | Lambda(id, t, bd) ->
+        check_no_nosimpl
+          (let open Context.Rel.Declaration in
+           push_rel (LocalAssum (id, t)) env) bd
+           (* todo: include also l in the arguments, but they probably
+              have to be "lifted" to be interpreted in the current
+              environment. *)
+      | _ -> true in
+  let rec srec env n labs onlyproj c : constant_evaluation =
     let c',l = whd_betadeltazeta_stack env sigma c in
```

Comment:
What I was suggesting is the one could pass more flags to this function, in particular 
https://github.com/coq/coq/blob/94d579844817edcbb2454dd9dc79071b2cd1d12a/kernel/cClosure.mli#L53
plus
https://github.com/coq/coq/blob/94d579844817edcbb2454dd9dc79071b2cd1d12a/kernel/cClosure.mli#L43
plus some code to fetch all key-value pairs such that value is "simpl never"
https://github.com/coq/coq/blob/94d579844817edcbb2454dd9dc79071b2cd1d12a/pretyping/reductionops.ml#L135
should let one force `whd_betadeltazeta_stack` to stop at any simpl-never flagged constants and one
can add a case in the code below raising `Redelimination` as you do in this patch.

It's a bit of code because one has to patch reductionops to "expose" the reduction flags.
IIRC @ppedrot was also interested to be able to pass these flags down to cClosure